feat: add image support to Cohere Chat Generator

ChinmayBansal opened the pull request:
### Related Issues

  - fixes #2126

  ### Proposed Changes:

  This PR adds multimodal (image + text) support to CohereChatGenerator, enabling the component to process both text and
  images in chat messages. The implementation follows established patterns from the AnthropicChatGenerator,
  LlamaCppChatGenerator, and WatsonxChatGenerator multimodal support.

  **Key Features Added:**
  - Image format validation for supported formats (PNG, JPEG, WEBP, GIF)
  - Proper message conversion to Cohere API format with base64 data URIs
  - Support for multimodal models like command-a-vision-07-2025
  - Role-based image restrictions (images only allowed in user messages)
  - Comprehensive error handling for unsupported formats and edge cases
  - Pre-validation of images before processing for better error flow

  **Implementation Details:**
  - Updated `_format_message()` method to handle multimodal content while preserving order
  - Added image format validation constants `ImageFormat` and `IMAGE_SUPPORTED_FORMATS`
  - Enhanced component docstring with detailed usage examples for multimodal scenarios
  - Added type ignore comments due to SDK compatibility requirements (see notes below)
  - Pre-validate all images upfront before content processing (following LlamaCpp pattern)
  - Updated minimum haystack-ai version to 2.17.1 for ImageContent support

  ### How did you test it?

  **Unit Tests:**
  - ✅ `test_format_message_with_image()` - Tests proper multimodal message conversion
  - ✅ `test_format_message_with_unsupported_mime_type()` - Tests error handling for unsupported formats
  - ✅ `test_format_message_with_none_mime_type()` - Tests edge case with None mime type
  - ✅ `test_format_message_image_in_non_user_message()` - Tests role-based restrictions
  - ✅ `test_multimodal_message_processing_with_mock()` - Tests end-to-end multimodal processing with mocked model
  - ✅ `test_supported_image_formats()` - Tests all supported formats (PNG, JPEG, WEBP, GIF)
  - ✅ `test_multiple_images_in_single_message()` - Tests multiple image support

  **Integration Tests:**
  - ✅ `test_live_run_multimodal()` - Tests live API calls with real Cohere multimodal models

  **Code Quality Verification:**
  - ✅ All linting checks pass: `hatch run fmt`
  - ✅ All type checking passes: `hatch run test:types`
  - ✅ All unit tests pass: `hatch run test:unit`

  **Manual Verification:**
  - Tested multimodal message creation and conversion
  - Verified proper error messages for validation failures
  - Confirmed Cohere API format compatibility with data URI structure

  ### Notes for the reviewer

  - The implementation closely follows the patterns established in AnthropicChatGenerator, LlamaCppChatGenerator, and
  WatsonxChatGenerator
  - Image validation uses the same error message format as other integrations for consistency
  - The `data:mime-type;base64,data` format with `image_url` structure is required by Cohere API for multimodal processing per [official docs](https://docs.cohere.com/docs/image-inputs)
  - Added comprehensive test coverage that matches and exceeds the patterns used in Anthropic, LlamaCpp, and Watson tests
  - All edge cases are properly handled including None mime types and role restrictions
  - **Type ignore usage**: Several `# type: ignore` comments are necessary due to SDK compatibility - Cohere's typed SDK
  conflicts with multimodal dict requirements, but this doesn't affect type safety elsewhere
  - Cohere supports more image formats (PNG, JPEG, WEBP, GIF) compared to Watson (JPEG, PNG only) but fewer than some other providers

  ### Checklist

  - [x] I have read the [contributors  guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
  - [x] I have updated the related issue with new insights and changes
  - [x] I added unit tests and updated the docstrings
  - [x] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

vblagoje reviewed (changes requested):
Looks solid, minor changes requested. I'll also try it out myself after next round.

ChinmayBansal commented on integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py:
```diff
@@ -113,8 +124,39 @@ def _format_message(
             msg = "A `ChatMessage` must contain at least one `TextContent`, `ToolCall`, or `ToolCallResult`."
```
Fixed

ChinmayBansal commented on integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py:
```diff
@@ -113,8 +124,39 @@ def _format_message(
+                        # Cohere expects base64 data URI format
+                        # See: https://docs.cohere.com/docs/image-inputs
+                        image_url = f"data:{part.mime_type};base64,{part.base64_image}"
+                        content_parts.append({"type": "image_url", "image_url": {"url": image_url}})  # type: ignore[dict-item]
```
Replaced normal dicts with custom TypedDict implementation

ChinmayBansal commented:
Hi @vblagoje,

I have addressed your feedback. The main issue with the type ignore is expecting two different formats. Using any would get rid of the type ignore but also reduce type safety.

All issues have been addressed. 

Could you review?
Thanks!

anakin87 commented:
I'll take a look...

ChinmayBansal commented:
Hi @anakin87,

Thanks for the merge. Just a little curious on the reformatting on the core logic as I tried to implement it close to other implementations I have done. The new changes simplify the logic, just wanted to know as I am trying to keep it consistent as I work through the multimodal support issues.
